reference/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,229 @@
+    */
+    const auto num_rows = mtx->get_size()[0];
+    const auto m_row_ptrs = mtx->get_const_row_ptrs();
+    const auto m_cols = mtx->get_const_col_idxs();
+    const auto m_vals = mtx->get_const_values();
+    auto i_row_ptrs = inverse_mtx->get_row_ptrs();
+    auto i_cols = inverse_mtx->get_col_idxs();
+    auto i_vals = inverse_mtx->get_values();
+
+    // expect mtx and inverse_mtx to have the same number of elems
+    const auto num_elems = mtx->get_num_stored_elements();
+    // Copy sparsity pattern of original into the inverse of L
+    std::copy_n(m_row_ptrs, num_rows + 1, i_row_ptrs);
+    std::copy_n(m_cols, num_elems, i_cols);
+
+    gko::Array<ValueType> rhs_array{exec};  // RHS for local trisystem
+    // memory for dense trisystem in column major:
+    gko::Array<ValueType> trisystem_array{exec};
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        const auto i_row_begin = i_row_ptrs[row];
+        const auto i_row_end = i_row_ptrs[row + 1];
+        const auto i_row_elems = i_row_end - i_row_begin;
+
+        trisystem_array.resize_and_reset(i_row_elems * i_row_elems);
+        auto trisystem = trisystem_array.get_data();
+        std::fill_n(trisystem, i_row_elems * i_row_elems, zero<ValueType>());
+
+        for (size_type i = 0; i < i_row_elems; ++i) {
+            const auto col = i_cols[i_row_begin + i];
+            const auto m_row_end = m_row_ptrs[col + 1];
+            auto m_row_ptr = m_row_ptrs[col];
+            auto i_row_ptr = i_row_begin;
+            // Values in `trisystem` will be stored in transposed order
+            // to prevent the need for an explicit transpose.
+            // Since `trisystem` stores in column major, this is equivalent
+            // to storing it in row-major and reading in column major.
+            auto idx = i * i_row_elems;
+            while (m_row_ptr < m_row_end && i_row_ptr < i_row_end) {
+                const auto sparsity_col = i_cols[i_row_ptr];
+                const auto m_col = m_cols[m_row_ptr];
+                if (sparsity_col == m_col) {
+                    trisystem[idx] = m_vals[m_row_ptr];
+                    ++m_row_ptr;
+                    ++i_row_ptr;
+                    ++idx;
+                } else if (m_col < sparsity_col) {
+                    ++m_row_ptr;
+                } else {  // element not present -> leave value at 0
+                    ++i_row_ptr;
+                    ++idx;
+                }
+            }
+        }
+
+        rhs_array.resize_and_reset(i_row_elems);
+        auto rhs = rhs_array.get_data();
+
+        trs_solve(i_row_elems, trisystem, rhs);
+
+        // Drop RHS as a row to memory (since that is the computed inverse)
+        for (size_type i = 0; i < i_row_elems; ++i) {
+            const auto new_val = rhs[i];
+            auto idx = i_row_begin + i;
+            // Check for non-finite elements which should not be copied over
+            if (isfinite(new_val)) {
```

Comment:
```suggestion
            if (gko::isfinite(new_val)) {
```
use `gko::` should solve the MSVC compile problem

Reply:
Thanks!
I forgot about that when I wrote this code...